src/protocol/prf_sharding/mod.rs

```diff
@@ -0,0 +1,711 @@
+    record_id: RecordId,
+    did_trigger_get_attributed: &Replicated<Gf2>,
+    trigger_value: &BitDecomposed<Replicated<Gf2>>,
+) -> Result<BitDecomposed<Replicated<Gf2>>, Error>
+where
+    C: UpgradedContext<Gf2, Share = Replicated<Gf2>>,
+{
+    Ok(BitDecomposed::new(
+        ctx.parallel_join(
+            trigger_value
+                .iter()
+                .zip(repeat(did_trigger_get_attributed))
+                .enumerate()
+                .map(|(i, (trigger_value_bit, did_trigger_get_attributed))| {
+                    let c = ctx.narrow(&BitOpStep::from(i));
+                    async move {
+                        trigger_value_bit
+                            .multiply(did_trigger_get_attributed, c, record_id)
+                            .await
+                    }
+                }),
+        )
+        .await?,
+    ))
+}
+
+///
+/// To provide a differential privacy guarantee, we need to bound the maximum contribution from any given user to some cap.
+///
+/// The following values are computed for each row:
+/// (1) The uncapped "Attributed trigger value" (which is either the original `trigger_value` bits or zero if it was unattributed)
+/// (2) The cumulative sum of "Attributed trigger value" thus far (which "saturates" at a given power of two as indicated by the `is_saturated` flag)
+/// (3) The "delta to cap", which is the difference between the "cap" and the cumulative sum (this value is meaningless once the cumulative sum is saturated)
+///
+/// To perfectly cap each user's contributions at precisely the cap, the "attributed trigger value" will sometimes need to be lowered,
+/// such that the total cumulative sum adds up to exactly the cap.
+///
+/// This oblivious algorithm computes the "capped attributed trigger value" in the following way:
+/// IF the cumulative is NOT YET saturated:
+///     - just return the attributed trigger value
+/// ELSE IF the cumulative sum JUST became saturated (that is, it was NOT saturated on the preceding line but IS on this line):
+///     - return the "delta to cap" from the preceding line
+/// ELSE
+///     - return zero
+///
+async fn compute_capped_trigger_value<C>(
+    ctx: C,
+    record_id: RecordId,
+    is_saturated: &Replicated<Gf2>,
+    is_saturated_and_prev_row_not_saturated: &Replicated<Gf2>,
+    prev_row_diff_to_cap: &BitDecomposed<Replicated<Gf2>>,
+    attributed_trigger_value: &BitDecomposed<Replicated<Gf2>>,
+) -> Result<BitDecomposed<Replicated<Gf2>>, Error>
+where
+    C: UpgradedContext<Gf2, Share = Replicated<Gf2>>,
+{
+    let narrowed_ctx1 = ctx.narrow(&Step::ComputedCappedAttributedTriggerValueNotSaturatedCase);
+    let narrowed_ctx2 = ctx.narrow(&Step::ComputedCappedAttributedTriggerValueJustSaturatedCase);
+
+    let one = &Replicated::share_known_value(&narrowed_ctx1, Gf2::ONE);
+
+    Ok(BitDecomposed::new(
+        ctx.parallel_join(
+            attributed_trigger_value
+                .iter()
+                .zip(prev_row_diff_to_cap.iter())
```

Comment:
```suggestion
            zip(attributed_trigger_value.iter(), prev_row_diff_to_cap.iter())
```
I tend to prefer this formulation.